### DISCUSSION TITLE
Why do carriage returns in source MD document result in hard line break?

### DISCUSSION THREAD

**Discussion by bertput** [vkbo/novelWriter#939] [Help]
Hello,

As I understand the markdown syntax, if two lines flow one after the other without a blank line in between, the text should flow in the rendered output or (I guess) the preview.  I think this is mentioned in another question but that turned into a discussion about MD vs. other formats.

Specifically, if I write some text like this:
```
these are some very
short lines
of text.
```
then I would expect it to render like this:
```
these are some very short lines of text.
```

Only an empty line should cause a hard line break, but I'm not seeing that.  Instead I see line breaks at the end of each line.
If there is a setting that I'm missing, I'd appreciate a pointer to it!  Thank you!

Regards,    Bert

**Comment by vkbo** [comment]
Hi Bert,

The novelWriter GUI is not intended to be a source/preview style setup. The editor renders the text as close as it can to the final result so you don't really need a preview. The view panel is simply there so you can view your notes while you write. This is the intended usage, and is the primary principle that informs how the editor and the syntax is designed. Because of this, line breaks are just line breaks. They should look the same in the editor and viewer, which your examples do not. The blank line is converted to a paragraph break, but that looks visually similar, and mostly only matters for the export formats.

Keep in mind that novelWriter it is not a markdown editor per say. It is a plain text editor with a limited markdown-like syntax for emphasis and headers, and a novelWriter-specific syntax for meta data and paragraph alignment.

I have considered dropping the markdown-like syntax altogether and allow the editor to hold rich text with limited formatting. This would make it less confusing to people who are used to plain text markup languages (I'm one as well). I initially just wanted to avoid the clutter of a rich text editor and go for simplicity. The app is only intended as a tool to organise notes and write a draft manuscript with a minimum amount of fluff.

**Reply by bertput** [reply]
Hi vkbo,

I have no problem with the editor or the previewer showing carriage returns, although I think the previewer flow the text better :-)  I can live with carriage returns.

However, the problem occurs when I build the project to, say, HTML.  The example above would render in HTML as:

```
these are some very<br/>
short lines<br/>
of text.</p>
```

In a browser, these render as very short lines of text.  I cannot find any way to modify that behavior.  I'd love to be able to throw a switch in preferences or something so that the built output can or cannot have those line breaks included. :-)

BTW I would very much prefer to keep to plain-text style content; I use git routinely to version control my projects so I'm all about plain-text formats. :-)

**Reply by vkbo** [reply]
> However, the problem occurs when I build the project to, say, HTML. The example above would render in HTML as:

Well, yes, as I said, the line breaks are preserved. That includes on export, otherwise it would be inconsistent and confusing. The viewer window is rendered from HTML4 anyway.

> In a browser, these render as very short lines of text. I cannot find any way to modify that behavior. I'd love to be able to throw a switch in preferences or something so that the built output can or cannot have those line breaks included. :-)

Why do you need to add them if you don't want them?

> BTW I would very much prefer to keep to plain-text style content; I use git routinely to version control my projects so I'm all about plain-text formats. :-)

I would still store the files in a format that would be suitable for version control. Either `.fodt`, or a similarly structured format using JSON instead of XML. The main reason to switch to a rich text format is that a lot of the feature requests can only be implemented in a user friendly way in a rich editor. Otherwise the syntax risks growing into something similar to LaTeX in complexity, and the readability is lost.

If I do switch to a rich text editor, it will be a version 2.0 release, and I would maintain a version 1.x LTS release with the current editor. I'm also working on a new application from scratch that builds on everything I've learned from making novelWriter, and incorporates a lot of the feature requests I've gotten over the years. If that one becomes viable, I will not implement these features as novelWriter 2.0.

**Reply by bertput** [reply]
> >  In a browser, these render as very short lines of text. I cannot find any way to modify that behavior. I'd love to be able to throw a switch in preferences or something so that the built output can or cannot have those line breaks included. :-)
> > 
>  Why do you need to add them if you don't want them?

Sometimes I will pull in text from other projects and it won't necessarily be formatted the same as the existing chapter.  That text would look quite different to the rest of the project when it is rendered.
I've used a few markdown editors and they generally allow you to specify if you want this behavior or not in the render or the preview.

Anyway, thanks for the opportunity to discuss it; good luck with the project!

**Reply by vkbo** [reply]
Well, I can look into adding it as an option. Please make a feature request. I don't have a lot of time to add new features right now, but I will get to it eventually. :)

**Reply by bertput** [reply]
Hmm, I think I fixed it with a one-line change:

```
diff --git a/novelwriter/core/tomd.py b/novelwriter/core/tomd.py
index 95218b7c..e714cbe2 100644
--- a/novelwriter/core/tomd.py
+++ b/novelwriter/core/tomd.py
@@ -108,7 +108,7 @@ class ToMarkdown(Tokenizer):
             # Process Text Type
             if tType == self.T_EMPTY:
                 if len(thisPar) > 0:
-                    tTemp = ("  \n".join(thisPar)).rstrip(" ")
+                    tTemp = ("\n".join(thisPar)).rstrip(" ")
                     tmpResult.append(f"{tTemp}\n\n")
                 thisPar = []
```

Maybe I should work on three hours sleep more often :-)

I'll leave it up to you how (ir if :-)) you want to integrate this change.  The problem seems to be the two spaces your markdown exporter puts at the end of every line.  I just removed those two spaces.

Have a good New Year!

**Reply by vkbo** [reply]
This is the wrong place for this change. It breaks the correct formatting of in-paragraph line breaks in markdown exports making it impossible to add such line breaks altogether. The change must be added to the processing of the T_TEXT block. It must also be applied to the HTML and ODT converter classes. These classes were originally written to do what you are requesting, so the code change you're asking for is already there in an earlier version of novelWriter (except for ODT I think). All that is needed is to restore the old code.

In any case, this behaviour has to be optional, and opt-in at that. So it needs to sit behind a flag that is either controlled from Preferences or directly in the Build tool. I think the latter is more appropriate.

I propose making a feature request. I can then pull up the old code, add it back in and add the switch. I plan to do a redesign of the build tool to allow for more options anyway, and this is a good feature to add into that mix. I also want to add line wrapping for Markdown exports.

As soon as I am finished with the current round of code cleanup, and standardisation, I want to start on a series of updates to the user interface based on user feedback. The build tool has been on my list for a while now.

In the mean time, have a happy new year, and thanks for your feedback :)

**Reply by bertput** [reply]
Yes, I knew it was a quick hacky change, which is why I'm fine if you don't include it.  I recognize that it fixes a very specific situation.  And yes, I agree that functionality should live behind a preferences flag or a build tool flag.

I'll create the feature request.  Much appreciated!
